[wasm] More jiterpreter cleanup

kg opened the pull request:
Blocks #78428, addresses some PR feedback and quality issues:
* Move more jiterpreter configuration and constants into options
* Fix should_generate_trace_here not scanning across multiple basic blocks
* Disable specialized JIT call in threaded wasm mode (though I think it might work, it's better to turn it off to be sure for now)
* Introduces genmintops.py, a script that automatically generates mintops.ts from mintops.def
* Adjust typescript config to make it able to find the generated mintops.ts (and fix ESLint on Linux)
* Unroll memsets below a certain size into raw wasm opcodes, because v8 generates expensive function calls for memset and memcpy. Unrolling memcpy is a TODO for later
* Rename "always generate" to "disable heuristic" to more accurately describe what it does
* Fix jiterpreter_dump_stats hiding errors if startup failed before cwraps were ready
* Misc. code cleanup

kg committed:
Fix problems with the jiterpreter heuristic's basic block traversal
Remove duplicated minimum opcode counter in the jiterpreter typescript
Move minimum jiterpreter trace hit count to options-def.h

kg committed:
Move more jiterp configuration into options, clean up a syntax warning

kg committed:
Fix eslint on linux

kg committed:
Automatically generate the typescript version of mintops.def at build time using a python script

lambdageek commented on src/mono/mono/mini/CMakeLists.txt:
```diff
@@ -518,13 +518,11 @@ add_custom_command(
-if (FALSE)
 add_custom_target(GENERATE_MINTOPS_TYPESCRIPT ALL
-  ${Python3_EXECUTABLE} ${CMAKE_CURRENT_SOURCE_DIR}/genmintops.py "${CMAKE_CURRENT_SOURCE_DIR}/interp/mintops.def" "${CMAKE_CURRENT_SOURCE_DIR}/../wasm/runtime/mintops.ts"
+  ${Python3_EXECUTABLE} ${CMAKE_CURRENT_SOURCE_DIR}/genmintops.py "${CMAKE_CURRENT_SOURCE_DIR}/interp/mintops.def" "${CMAKE_CURRENT_SOURCE_DIR}/../../wasm/runtime/mintops.ts"
```
@kg I kind of want to experiment with generating this file into `artifacts/` and then using [TypeScript's path mapping](https://www.typescriptlang.org/docs/handbook/module-resolution.html#path-mapping) or maybe its `rootDirs` config setting (same link, further down) to find the `dotnet-genrated/mintops.ts` file.  (ie add a path mapping for `dotnet-generated` to `$(REPO_TOP)/artifacts/blah/blah/blah`).

I'll try to pay around this weekend.  Not 100% confident it'll work with rollup+TS. but maybe...

lambdageek replied:
It's kind of working, in principle.  Except in practice I'm getting some weird unrelated error from rollup.  (it didn't load the typescript plugin and thinks `exports.ts` has some invalid JS syntax)

One observation, @kg that I think it true even for your current version of this stuff:  Running `genmintops.py` from cmake is not really necessary, and is kind of wrong:  the cmake build runs _after_ the rollup TS->JS build.

Instead I think we should just run the python script from MSBuild.

I added this to my wasm.proj instead of the cmake custom command:

```xml
  <Target Name="GenerateMintopsTS"
          Inputs="$(MonoProjectRoot)wasm\runtime\genmintops.py;$(MonoProjectRoot)mono\mini\interp\mintops.def"
          Outputs="$(NativeGeneratedFilesDir)\mintops.ts">
      <Exec Command="$(PythonCmd) $(MonoProjectRoot)wasm/runtime/genmintops.py $(MonoProjectRoot)mono/mini/interp/mintops.def $(NativeGeneratedFilesDir)/mintops.ts" />
  </Target>
```

And then added `GenerateMintopsTS` to the `BuildWithRollup` target:

```xml
  <Target Name="BuildWithRollup"
          Inputs="@(_RollupInputs);$(_RollupEnvIntermediateFile)"
          Outputs="$(NativeBinDir).rollup-stamp"
	  DependsOnTargets="WriteRollupEnvToFile;GenerateMintopsTS"
          >
```

kg replied:
Will try migrating to msbuild, thanks.

kg committed:
Checkpoint raytracer cleanups

kg committed:
Checkpoint mintops ts generation changes

kg committed:
Unroll small memsets because v8's codegen for memory.fill is awful

kg committed:
Checkpoint: Move genmintops

kg committed:
Checkpoint

kg committed:
Move dumpTraces into options.h. Don't update mintops.ts unless the contents need to change.

kg committed:
Fix error being swallowed if a failure happens during startup

kg committed:
Address PR feedback

lambdageek committed:
generate mintops.ts to artifacts/bin/native/generated

- Drive the generation from wasm.proj (driving it from cmake is wrong,
because the cmake build runs after rollup)
- Use the tsconfig rootDirs option to specify where to find the
files (runtime dir and also artifacts/bin/native/generated)
- Also due to rollup issues, specify rootDirs and include in
rollup.config.js - otherwise the rollup typescript plugin doesn't seem
to kick in, and barebones rollup gets confused when parsing .ts files
- The outDir option in tsconfig.json had to be deleted to make rollup
work.  But it was also unused since rollup controls where the ts
plugin is supposed to emit its output
- VS Code understands the tsconfig.json changes and can follow
references to the artifacts/.../generated folder

Note that Release and Debug builds both dump the generates mintops.ts
in the same folder.  This seems ok for now since there's no special
debug-only opcodes.  (Also VS Code has no idea about configurations,
so it has to be done this way or else it won't be able to find the .ts file)

lambdageek commented:
@kg please cherrypick https://github.com/lambdageek/runtime/commit/91d2bc90758eb188f06ea1893f257606db818706

That will reconfigure the build to:
1. Run `genmintops.py` from `wasm.proj` instead of cmake.  This is necessary because cmake runs after rollup, but rollup needs the .ts file.
2. Place the generated `mintops.ts` file into `artifacts/bin/native/generated`

There's a bunch of required rollup and tsconfig changes in there that work around some kind of bug where rollup fails to delegate to typescript when a certain combination of options is used.

As far as I can tell VS Code is happy with these changes - after a build, chasing references in VS Code works - it will find the opcode definitions in the artifacts folder and warn appropriately.

kg committed:
Checkpoint

kg committed:
Repair merge damage

kg committed:
Repair merge damage

kg committed:
Fix 3 byte memset

lambdageek commented on src/mono/wasm/wasm.proj:
```diff
@@ -406,10 +413,16 @@
 
+  <Target Name="GenerateMintopsTS"
+          Inputs="$(MonoProjectRoot)wasm\runtime\genmintops.py;$(MonoProjectRoot)mono\mini\interp\mintops.def"
+          Outputs="$(NativeGeneratedFilesDir)\mintops.ts">
+      <Exec Command="$(PythonCmd) $(MonoProjectRoot)wasm/runtime/genmintops.py $(MonoProjectRoot)mono/mini/interp/mintops.def $(NativeGeneratedFilesDir)/mintops.ts" />
```
```suggestion
          Outputs="$(NativeGeneratedFilesDir)mintops.ts">
      <Exec Command="$(PythonCmd) $(MonoProjectRoot)wasm/runtime/genmintops.py $(MonoProjectRoot)mono/mini/interp/mintops.def $(NativeGeneratedFilesDir)mintops.ts" />
```

lambdageek reviewed (commented):
Some MSBuild nits; and an explicit dependency on mintops.ts for the main rollup target

kg committed:
Update src/mono/wasm/wasm.proj

Co-authored-by: Aleksey Kliger (λgeek) <akliger@gmail.com>

kg committed:
Update src/mono/wasm/wasm.proj

Co-authored-by: Aleksey Kliger (λgeek) <akliger@gmail.com>

kg committed:
Add missing always_inline

kg committed:
Merge remote-tracking branch 'origin/wasm-jiterpreter-cleanup-2' into wasm-jiterpreter-cleanup-2

kg committed:
Fix threading define

kg committed:
Disable jiterpreter if tiering is disabled

kg committed:
Address PR feedback

vargaz commented:
The mono changes look ok to me.

pavelsavara commented:
>     2. Place the generated `mintops.ts` file into `artifacts/bin/native/generated`

I don't agree that it should be missing in git. 
It breaks intelisense until you re-compile.

Perhaps I don't understand all of it.
Why do you think it's good idea @lambdageek ?

lambdageek commented:
> >     2. Place the generated `mintops.ts` file into `artifacts/bin/native/generated`
> 
> I don't agree that it should be missing in git. 
> It breaks intelisense until you re-compile.
> 
> Perhaps I don't understand all of it.
> Why do you think it's good idea @lambdageek ?
> 
> 

Checking in a generated file makes workflow more complicated:

1. For the inner dev loop it's harder to get incremental builds right.
2. For someone working in the interpreter in another platform it's not obvious that they need to regenerate a wasm file.

As for intellisense, in vs code I only get errors without rebuilding if I open jiterpreter.ts. opening other files works fine.
